### PR TITLE
#1636: New parser for AdultSearch

### PR DESCRIPTION
**Pull request by lucywhg1**
EDIT: All done!
Currently have most of adparser implemented, with the exception of matching for phone numbers and emails. Tests are running right but the assertions don't fully match yet because I haven't filled out the ad text.

### CODE REVIEW

File: __app__/adparser/sites/adultsearch_com.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        category = self.breadcrumb[-1]
+
+        if "Massage" in category or self.about_section.get("Massage"):
+            services.append("massage")
+
+        if "Escort" in category:
+            services.append("escort")
+
+        if self.about_section.get("Fetish Session"):
+            services.append("bdsm")
```
Comment (lizrush) [inline __app__/adparser/sites/adultsearch_com.py:83]: interesting approach to grab these pieces of info from different sections to get a robust view of what the services are! 👍 :

File: __app__/adparser/sites/base_ad_parser.py [hunk 2]
```diff
@@ -86,6 +87,6 @@ def ad_dict(self) -> dict:
             "website": self.website(),
             "ad-text": self.ad_text(),
             "ad-title": self.ad_title(),
-            "orientation": self.orientation()
+            "orientation": self.orientation(),
```
Comment (lizrush) [inline __app__/adparser/sites/base_ad_parser.py:90]: i think the trailing comma shouldn't be here. Do you have the auto-formatter Black for python installed in Vscode? that might be helpful!

File: .gitattributes [hunk 3]
```diff
@@ -0,0 +1 @@
+* text=auto
```
[the remark this answers is not in the document]
Comment (lucywhg1) [reply .gitattributes:1]: Windows and Linux use different line endings for their files, so whenever there's a team with mixed OS's, all the Windows folks (or all the Linux folks, I suppose) have to manually set line endings to convert in the repo. I was doing that for a while, but there was a bit of a mix-up recently that necessitated a bunch of back and forth merging with line endings, so I looked it up and this addition to .gitattributes is the widespread solution. 

It just standardizes line endings ("auto" conversion) in the repo so nothing breaks for anyone.

File: __app__/adparser/sites/adultsearch_com.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    def date_posted(self) -> str:
+        return None
+
+    def name(self) -> str:
+        name = self.soup.select_one(".details__heading").find("span")
+        return name.text.strip()
+
+    def primary_email(self) -> str:
+        return self.about_section.get("Email")  # did not see any ads with Email here
```
[the remark this answers is not in the document]
Comment (lucywhg1) [reply __app__/adparser/sites/adultsearch_com.py:32]: There is no section that I ever saw, but there have been labels for Phone: and Website:, so I figure it would probably be under a similar listing if it did exist, and if not, then when I add in the `emails` method that should catch it elsewhere in the ad.

File: __app__/adparser/sites/base_ad_parser.py (hunk 2 above)
Comment (lucywhg1) [reply __app__/adparser/sites/base_ad_parser.py:90]: Looks like Black is installed, but I'll pull this out.

File: __app__/adparser/sites/adultsearch_com.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        return None
+        # not yet implemented in this PR
+
+    def social(self) -> List:
+        return None
+
+    def age(self) -> str:
+        stats = self.about_section.get("Stats").split(", ")
+        if len(stats) > 0:
+            print("0 stats is " + str(stats[0].strip().replace(" years old", "")))
```
[the remark this answers is not in the document]
Comment (lucywhg1) [reply __app__/adparser/sites/adultsearch_com.py:44]: A few of them still lingering, will take them out once I wrap up the last touch of adparser!

File: __app__/adlistingparser/tests/test_ad_listing_parsers.py [hunk 6]
```diff
@@ -23,6 +23,7 @@ def case_data(case_loc: str) -> List:
 
     return test_cases
 
+
```
[the remark this answers is not in the document]
Comment (lucywhg1) [reply __app__/adlistingparser/tests/test_ad_listing_parsers.py:26]: Again I think this may be black formatting with double newlines after methods. Maybe later when we poke through the crawlers we can double-check my VSCode setup so I'm not running around reformatting things on-save accidentally :P

File: __app__/adlistingparser/adlistingparser.py [hunk 7]
```diff
@@ -2,6 +2,8 @@
 from typing import List
 import logging
 from copy import deepcopy
+
+from __app__.adlistingparser.sites.adultlook_com import AdultLook_com
```
[the remark this answers is not in the document]
Comment (lucywhg1) [reply __app__/adlistingparser/adlistingparser.py:6]: Agree with the nit! Done.

File: __app__/sitemapparser/sites/adultsearch_com.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+from urllib.parse import urljoin, urlparse
+from bs4 import BeautifulSoup
+import requests
+import sys
+
+
+# Catch regional subdomains.
```
Comment (lizrush) [inline __app__/sitemapparser/sites/adultsearch_com.py:7]: I think it would be nice to expand on this comment with an example and explanation of why we are getting this regional subdomain.

File: __app__/sitemapparser/sites/adultsearch_com.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+def _find_domain(soup):
+    uri = urlparse(
+        soup.find("link", rel="canonical").get("href")
+    )  # captures country subdomains
+
+    return f"{uri.scheme}://{uri.netloc}"
+
+
+# Attempt to return the soup of a tag's linked page.
```
Comment (lizrush) [inline __app__/sitemapparser/sites/adultsearch_com.py:16]: What does `tag` mean in this context?

File: __app__/sitemapparser/sites/adultsearch_com.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    return region_soups
+
+
+# Retrieves ad listings from nested sitemap>region>category structure.
+def adultsearch_com(html):
+    sitemap_soup = BeautifulSoup(html, "html.parser")
+    ad_listing_links = []
+
+    region_soups = get_region_soups(sitemap_soup)
```
Comment (lizrush) [inline __app__/sitemapparser/sites/adultsearch_com.py:60]: I think we could rename things to avoid "soups", which isn't really obvious what that is (unless you are already familiar with beautiful soup). Why not just `get_regions` instead of `get_region_soups`?

File: __app__/siteloader/siteloader.py [hunk 11]
```diff
@@ -10,6 +10,7 @@
     ("megapersonals.eu", {"domain": "megapersonals.eu"}),
     ("escortdirectory.com", {"domain": "escortdirectory.com"}),
     ("2backpage.com", {"domain": "2backpage.com"}),
+    ("adultsearch.com", {"domain": "adultsearch.com"}),
```
[the remark this answers is not in the document]
Comment (lucywhg1) [reply __app__/siteloader/siteloader.py:13]: Yeah, wasn't sure on this, might still take it out
https://media2.giphy.com/media/TiPQa86PpufKymqPAu/source.gif

File: __app__/sitemapparser/sites/adultsearch_com.py (hunk 10 above)
Comment (lucywhg1) [reply __app__/sitemapparser/sites/adultsearch_com.py:60]: That makes sense to me - since I already use it in a couple pages, I think I'll go with pages, just to be clear that we're not getting a list of regions or anything but actual HTML.

File: __app__/sitemapparser/sites/adultsearch_com.py (hunk 9 above)
Comment (lucywhg1) [reply __app__/sitemapparser/sites/adultsearch_com.py:16]: Hmm, I think I really mean a link (an <a> tag), but since href is a different word, I'll switch this to saying links.

File: __app__/sitemapparser/sites/adultsearch_com.py (hunk 8 above)
Comment (lucywhg1) [reply __app__/sitemapparser/sites/adultsearch_com.py:7]: Agreed!

File: __app__/adparser/sites/adultsearch_com.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+    def _get_breadcrumb(self) -> List:
+        # Home > Country > State (for CA/US) > Locality > Category > Phone #
+        breadcrumb_text = [
+            crumb.text.strip()
+            for crumb in self.soup.select_one(".breadcrumb").find_all("li")
+        ]
+        phone_index = len(breadcrumb_text) - 1  # return list with category as last elem
+
+        return breadcrumb_text[1:phone_index]  # exclusive end
```
[the remark this answers is not in the document]
Comment (lucywhg1) [reply __app__/adparser/sites/adultsearch_com.py]: Yes, that's what I mean, except that I'm excluding _both_ Home and the Phone Number, so not just popping last (essentially I'm stripping off both ends). I updated the comment to be clearer, but let me know if you think there should be additional changes.

[the remark this answers is not in the document]
Comment (lizrush) [reply __app__/adparser/sites/adultsearch_com.py]: Got it, got it. No that makes plenty of sense! I didn't catch the first part about starting at the 1 index!

File: __app__/adlistingparser/sites/adultsearch_com.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+    def continuation_url(self):
+        maybe_next_page = self.soup.select_one(".next-arrow")
+        if maybe_next_page:
+            return maybe_next_page.get("href")  # contains full url + domain
+
+    def ad_listing_data(self) -> dict:
+        # adlisting does not hold location or services, on per ad basis
+        metadata = {
+            "gender": "female"  # almost all ads are female; ad parser catches exceptions
```
Comment (peoplespete) [inline __app__/adlistingparser/sites/adultsearch_com.py:27]: I'm a little torn about this. My instinct is if there isn't specific gender data on the page, let's not fill it in.

That said, do we do this sort of thing on other crawlers? 
That would convince me this is fine.

File: __app__/adlistingparser/tests/test-html/20200825_adultsearch_com_1.html [hunk 14]
```diff
@@ -0,0 +1,2459 @@
+<!DOCTYPE html>
+<html>
```
Comment (peoplespete) [inline __app__/adlistingparser/tests/test-html/20200825_adultsearch_com_1.html:2]: If we are doing 3 files for the adlistingparser test html, let's add that to the documentation for writing a new parser in case it isn't already there.

File: __app__/adparser/sites/adultsearch_com.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        if len(stats) > 0:
+            return stats[0].strip().replace(" years old", "")
+
+    def image_urls(self) -> List:
+        images_carousel = self.soup.select_one("#ad").select_one(".carousel")
+
+        return [("https:" + img["src"]) for img in images_carousel.find_all("img")]
+
+    def location(self) -> str:
+        loc_info = self.breadcrumb[:-1]  # all but last category crumb
```
Comment (peoplespete) [inline __app__/adparser/sites/adultsearch_com.py:64]: 🥇

File: __app__/adparser/sites/adultsearch_com.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+    def ad_title(self) -> str:
+        heading_items = self.soup.select_one(".details__heading").find_all("span")
+        return heading_items[1].text.strip()
+
+    def orientation(self) -> str:
+        return self.about_section.get("Available To")
+
+    # Member methods
+    def _get_about_section(self) -> Dict:  # contains stats, services
```
Comment (peoplespete) [inline __app__/adparser/sites/adultsearch_com.py:121]: This is super clear. If you have a minute, please add a little explanation of this standard to the docs (# Member methods + starting them with _).

It seems like we are very inconsistent in how we have done this in other crawlers and the way you are doing it here strikes me as the best I've seen so I'd love to essentially say, this is how you do it. Then when we make or review others, we can say, do it like this one.

File: __app__/sitemapparser/sites/adultsearch_com.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    uri = urlparse(soup.find("link", rel="canonical").get("href"))
+
+    return f"{uri.scheme}://{uri.netloc}"
+
+
+# Attempt to retrieve the linked page.
+def _get_linked_page(link, domain):
+    resp = requests.get(urljoin(domain, link.get("href")))
+
+    if resp:
```
Comment (peoplespete) [inline __app__/sitemapparser/sites/adultsearch_com.py:19]: Let's move these methods to the bottom under a `# Member methods` comment, to follow the convention above.

File: __app__/sitemapparser/tests/requests_mocks.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+    def __init__(self):
+        # Only return 1 province, state, and linked category pages for dozens of links.
+        self.should_return = {
+            "region": {"us": True, "ca": True},
+            "category": {"us": True, "ca": True},
+        }
+
+    def _get_resp_for(self, url, key):
+        if self.should_return[key]["ca"] and url.netloc == "ca.adultsearch.com":
```
Comment (peoplespete) [inline __app__/sitemapparser/tests/requests_mocks.py:33]: Under what circumstances would this be false?
`self.should_return[key]["ca"]`

File: docs/network-request-parsers.md [hunk 19]
```diff
@@ -0,0 +1,46 @@
+## Sitemap Parsers Using Network Requests
+
+For some sitemaps, it may not be possible to directly retrieve ad listing urls by grabbing links on the page or appending category and locations to the domain. For example, *adultsearch.com* does not have a search bar (which, as with *adultlook.com*, can be queried with an empty string for an informal adlisting page), and adlistings are stored in a nested structure.
+<img src="C:\Users\Lucy\Documents\GitHub\crawlers\docs\imgs\nested_sitemap.JPG" alt="nested_sitemap" style="zoom: 33%;" />
+
+As shown (blurrily) above, the structure of *adultsearch.com* goes **sitemap>region>category>ad listings by city**. In order to retrieve each of these subsequent pages, `app/sitemapparser/sites/adultsearch.py` uses the [requests](https://requests.readthedocs.io/en/master/) module. At each level, it finds a link to the next page (e.g. on the landing sitemap page, it selects all `h5 >a` tags, which are regional links like "Alabama"), makes a `requests.get(url)` call, and then parses the response using BeautifulSoup.
```
Comment (peoplespete) [inline docs/network-request-parsers.md:6]: I like this, I just want to make sure I understand.

So adding `requests` allows us to do this inside of our existing parsers instead of creating new steps in the ad processing and new queues in Azure. Am I understanding that correctly?

File: __app__/sitemapparser/tests/requests_mocks.py (hunk 18 above)
Comment (lucywhg1) [reply __app__/sitemapparser/tests/requests_mocks.py:33]: A touch further down in the file, should_return is set to false as soon as the page has been returned - that's because the Adult Search sitemap has 50 states, 15+ provinces, etc. and I don't want to be sending 50 duplicate pages back to the test parser. Instead, I just sent one sample page for each, and since the adultsearch sitemapparser will skip over any `requests.get` returns that are `None`, it basically trims down the tests immensely.

So basically `should_return` ensures our mocker sends back one US region page, one Canadian region page, one US category page, and one Canadian category page.

File: docs/network-request-parsers.md (hunk 19 above)
Comment (lucywhg1) [reply docs/network-request-parsers.md:6]: Essentially, yes - requests allows us to navigate through the website to get to the adlistings, which is _still_ 'site map parsing,' it's just slightly more complex parsing than what we've done for past sites (and in most cases, ideally would continue to do), which is simply grabbing all the adlisting links from the front page or constructing them from stuff we find on that first page.

File: __app__/adlistingparser/sites/adultsearch_com.py (hunk 13 above)
Comment (lucywhg1) [reply __app__/adlistingparser/sites/adultsearch_com.py:27]: We do this on other crawlers so long as it's pretty clear from browsing that the vast majority of posts are female _and_ that exceptions are called out by category/tag/etc. (for example, AdultSearch specifically has "shemale" or "male to male" categories if an ad is not for someone who identifies as female).

File: __app__/adlistingparser/tests/test-html/20200825_adultsearch_com_1.html (hunk 14 above)
Comment (lucywhg1) [reply __app__/adlistingparser/tests/test-html/20200825_adultsearch_com_1.html:2]: Done, and called it out a little more clearly.

File: __app__/sitemapparser/tests/requests_mocks.py (hunk 18 above)
Comment (peoplespete) [reply __app__/sitemapparser/tests/requests_mocks.py:33]: I think this can be greatly simplified by using MagicMock, our python mocking library.
`#side_effect` allows you to return a different value each time a mock is called.

This file has one:
`./tests/adlistingparser/test_adlistingparser.py`

I'm new at all this as well but here are some examples:
https://aaronlelevier.github.io/python-unit-testing-with-magicmock/
Section: `Mock multiple return values`

Also if we need conditional return values:
  https://stackoverflow.com/questions/29562460/conditional-mocking-call-original-function-if-condition-does-match/29563665
  Michele d'Amico's answer

Comment (lucywhg1) [reply __app__/sitemapparser/tests/requests_mocks.py:33]: Okay, so Pete and I talked about this some, and what's currently been submitted does not use MagicMock. My reasoning for that is that `setattr` on monkeypatch is already another form of mocking (whatever you set the value to _should be_ your 'mocked' method/test case implementation). Instead, it seemed like the benefit of MagicMock Pete was listing above was that it could return multiple values such that the first is our HTML page and all subsequent calls return None.

However, I think that functionality is really more about iterators/generators than MagicMock itself, and I didn't want to define a near identical generator for every page we wanted to only return once. So what I did is created something that can (and I think _will_ be used in the future, which is why I felt comfortable making a short class for it) return a mocked HTML page the first time and ignores all future calls. It looks like this:
```
class RespondOnce:
    def __init__(self, filename):
        self.page = MockResponse(get_html_text(filename))
        self.generator = self.gen()

    def gen(self):
        yield self.page
        while True:
            yield None  # skip response for subsequent calls

    @property
    def resp(self):
        return next(self.generator)
```
And is used like:
```
self.us_region = RespondOnce("20200827_adultsearch_com-region-alabama.html")
...
return self.us_region.resp
```

I think this makes the mocking more understandable than the previous dict implementation I had, but let me know. It just seems to me that if all we need from MagicMock is the ability to grab one value then skip, we're really looking at a lazy iterator.
